rpc/remote.go

```diff
@@ -0,0 +1,170 @@
+	// returns an address as if it were called the first time.
+	Reset()
+	// String returns a string that represents all addresses in Remote and can
+	// be used to construct a Remote that behaves the same way.
+	String() string
+}
+
+type fixedRemote string
+
+// NewFixedRemote returns a remote that always uses remoteAddr.
+func NewFixedRemote(remoteAddr string) Remote {
+	return fixedRemote(remoteAddr)
+}
+
+// GetAddress implements the Remote interface.
+func (r fixedRemote) GetAddress() string {
+	return string(r)
+}
+
+// Peek implements the Remote interface.
+func (r fixedRemote) Peek() string {
+	return string(r)
+}
+
+// Reset implements the Remote interface.
+func (r fixedRemote) Reset() {}
+
+// String implements the Remote interface.
+func (r fixedRemote) String() string {
+	return string(r)
+}
+
+type prioritizedRoundRobinRemote struct {
+	addresses [][]string
+
+	lock      sync.Mutex
+	toIterate [][]string
+}
+
+// NewPrioritizedRoundRobinRemote creates a new Remote that include
+// prioritized remote groups. Each call to GetAddress() will round-robin by
+// random order within the first group. If we run out of address within the
+// first group, fallback to second group and do the samething, until we've
+// iterated all groups where we'll start over from first group.
+//
+// Any successful connecting attempt should result in a call to Reset(). This
+// is generatlly handled by the rpc package itself and shouldn't be worried by
+// the user of rpc package unless noted otherwise.
+func NewPrioritizedRoundRobinRemote(addressGroups [][]string) (Remote, error) {
+	// filter out empty ones
+	cleaned := make([][]string, 0, len(addressGroups))
+	for _, group := range addressGroups {
+		cleanedGroup := make([]string, 0, len(group))
+		for _, addr := range group {
+			addr := strings.ToLower(strings.TrimSpace(addr))
+			if len(addr) > 0 {
+				cleanedGroup = append(cleanedGroup, addr)
+			}
+		}
+		if len(cleanedGroup) > 0 {
+			cleaned = append(cleaned, cleanedGroup)
+		}
+	}
+
+	if len(cleaned) == 0 {
+		return nil, errors.New("addressGroups has no address")
+	}
+
+	r := &prioritizedRoundRobinRemote{
+		addresses: cleaned,
+	}
+	r.resetLocked()
+	return r, nil
+}
+
+func (r *prioritizedRoundRobinRemote) resetLocked() {
+	r.toIterate = make([][]string, 0, len(r.addresses))
+	for _, group := range r.addresses {
+		groupCopied := make([]string, 0, len(group))
+		for _, i := range rand.Perm(len(group)) {
+			groupCopied = append(groupCopied, group[i])
+		}
+		r.toIterate = append(r.toIterate, groupCopied)
+	}
+}
+
+// Reset implements the Remote interface.
+func (r *prioritizedRoundRobinRemote) Reset() {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+
+	r.resetLocked()
+}
+
+// GetAddress implements the Remote interface.
+func (r *prioritizedRoundRobinRemote) GetAddress() string {
+	r.lock.Lock()
+	defer r.lock.Unlock() // If we have run out of addresses, reset to include all addresses and
```

Comment:
This comment is strangely placed.  Is it supposed to start on the next line?

Reply:
Ah weird. Probably an accidental Shift+j. Thanks!